pydeseq2/utils.py

```diff
@@ -157,152 +155,236 @@ def build_design_matrix(
 
-            # Check that the reference level is not in the matrix (if unexpanded design)
-            ref_level_name = "_".join(ref_level)
-            if (not expanded) and ref_level_name in design_matrix.columns:
-                # Remove the reference level and add one
-                factor_cols = [
-                    col for col in design_matrix.columns if col.startswith(ref_level[0])
-                ]
-                missing_level = next(
-                    level
-                    for level in np.unique(metadata[ref_level[0]])
-                    if f"{ref_level[0]}_{level}" not in design_matrix.columns
-                )
-                design_matrix[f"{ref_level[0]}_{missing_level}"] = 1 - design_matrix[
-                    factor_cols
-                ].sum(1)
-                design_matrix.drop(ref_level_name, axis="columns", inplace=True)
-
-        if not expanded:
-            # Add reference level as column name suffix
-            for factor in design_factors:
-                if ref_level is None or factor != ref_level[0]:
-                    # The reference is the unique level that is no longer there
-                    ref = next(
-                        level
-                        for level in np.unique(metadata[factor])
-                        if f"{factor}_{level}" not in design_matrix.columns
+    Raises
+    ------
+    ValueError
+        If design_factors is not a list.
+    ValueError
+        If a factor has only one level.
+    KeyError
+        If the reference level does not contain two strings.
+    KeyError
+        If the reference level is not in the metadata.
+    """
+    # Because of preprocessing we did in dds continuous and categorical variables
+    # are tagged in the metadata
+    # We just need special treatments if there are ref_levels
+    patterns_to_reverse = {}
+    if ref_level is not None:
+        assert isinstance(ref_level, list), "The reference level should be a list."
+        assert all(
+            isinstance(ref, tuple) for ref in ref_level
+        ), "The reference levels is not a list of tuples"
+        assert all(
+            len(ref) == 2 for ref in ref_level
+        ), "The reference levels should all contain 2 strings."
+        assert all(
+            ref[0] in metadata.columns for ref in ref_level
+        ), "The reference levels should all be in the metadata."
+        # This is how formulaic expects ref_level [("condition", "A")]:
+        # ~ C(condition, contr.treatment(base="A"))
+        for ref in ref_level:
+            assert ref[0] in set(
+                metadata.columns
+            ), "The reference level is not in the metadata"
+            assert (
+                ref[1] in metadata[ref[0]].unique()
+            ), f"The reference level is not in the metadata col={ref[0]}"
+            replacement_key = "C(" + ref[0] + ", contr.treatment(base='" + ref[1] + "'))"
+            patterns_to_reverse[replacement_key] = ref[0]
```

Comment:
This information is already available from formulaic directly. When `design` is your design matrix, there is
```
design.model_spec.variable_terms
```
and 
```
design.model_spec.term_variables
```

to map from variables to formulae terms and vice versa.

Reply:
Didn't know ! Thanks !

Reply:
I opened a corresponding issue